installer-downloader/src/winapi_impl/delegate.rs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    fn hide_download_progress(&mut self) {
+        self.progress_bar.set_visible(false);
+    }
+
+    fn set_download_progress(&mut self, complete: u32) {
+        self.progress_bar.set_pos(complete);
+    }
+
+    fn show_download_button(&mut self) {
+        self.download_button.set_visible(true);
+    }
+
+    fn hide_download_button(&mut self) {
+        self.download_button.set_visible(false);
+    }
+
+    fn enable_download_button(&mut self) {
+        self.download_button.set_enabled(true);
+    }
+
+    fn disable_download_button(&mut self) {
+        self.download_button.set_enabled(false);
+    }
+
+    fn show_cancel_button(&mut self) {
+        self.cancel_button.set_visible(true);
+    }
+
+    fn hide_cancel_button(&mut self) {
+        self.cancel_button.set_visible(false);
+    }
+
+    fn enable_cancel_button(&mut self) {
+        self.cancel_button.set_enabled(true);
+    }
+
+    fn disable_cancel_button(&mut self) {
+        self.cancel_button.set_enabled(false);
+    }
+
+    fn show_beta_text(&mut self) {
+        self.beta_prefix.set_visible(true);
+        self.beta_link.set_visible(true);
+    }
+
+    fn hide_beta_text(&mut self) {
+        self.beta_prefix.set_visible(false);
+        self.beta_link.set_visible(false);
+    }
+
+    fn queue(&self) -> Self::Queue {
+        Queue {
+            main_wnd: self.window.handle,
+        }
+    }
+}
+
+/// Register a window message for clicking this button that triggers `callback`.
+fn register_click_handler(
+    parent: nwg::ControlHandle,
+    button: nwg::ControlHandle,
+    callback: impl Fn() + 'static,
+) {
+    nwg::bind_event_handler(&button, &parent, move |evt, _, handle| {
+        if evt == Event::OnButtonClick && handle == button {
+            callback();
+        }
+    });
+}
+
+/// Queue sends a window message to the main window containing a [QueueContext], giving us mutable
+/// access to the [AppDelegate] on the main UI thread.
+///
+/// See [QueueContext] docs for more information.
+#[derive(Clone)]
+pub struct Queue {
+    main_wnd: nwg::ControlHandle,
+}
+
+// SAFETY: It is safe to post window messages across threads
+unsafe impl Send for Queue {}
+
+/// The context contains a callback function that is passed as a pointer to the main thread
+/// along with a custom window message `QUEUE_MESSAGE`.
+///
+/// It must be wrapped in a struct since we cannot pass a fat pointer
+/// `*mut dyn for<'a> FnOnce(&'a mut AppWindow) + Send` to `PostMessageW`.
+pub struct QueueContext {
+    pub callback: Box<dyn for<'a> FnOnce(&'a mut AppWindow) + Send>,
+}
+
+impl AppDelegateQueue<AppWindow> for Queue {
+    fn queue_main<F: FnOnce(&mut AppWindow) + 'static + Send>(&self, callback: F) {
+        let Some(hwnd) = self.main_wnd.hwnd() else {
+            return;
+        };
+        let context = QueueContext {
+            callback: Box::new(callback),
+        };
+        let context_ptr = Box::into_raw(Box::new(context));
```

Comment:
Could we get rid of one level of boxing by doing this?
```suggestion
        let context_ptr = Box::into_raw(Box::new(callback));
```

Reply:
No, `Box<dyn for<'a> FnOnce(&'a mut AppWindow) + Send>` stores a fat pointer, so it is too large.

Reply:
Ah, I see.